File: 02-download-a-file/ruby/download.rb
require 'selenium-webdriver'
require 'rspec/expectations'
require 'uuid'
require 'fileutils'

include RSpec::Matchers

def setup
  @download_dir = File.join(Dir.pwd, UUID.new.generate)
  FileUtils.mkdir_p(@download_dir)

  profile = Selenium::WebDriver::Firefox::Profile.new
  profile['browser.download.dir'] = @download_dir
  profile['browser.download.folderList'] = 2 # the last folder specified for download
  profile['browser.helperApps.neverAsk.saveToDisk'] = 'image/jpeg, application/pdf, application/octet-stream'
  profile['pdfjs.disabled'] = true # need to set with PDFs or else it will view them in the browser
  options = Selenium::WebDriver::Firefox::Options.new
  options.profile = profile
  @driver = Selenium::WebDriver.for(:firefox, options: options)
end

def teardown
  @driver.quit
  FileUtils.rm_rf(@download_dir)
end

def run
  setup
  yield
  teardown
end

run do
  @driver.get('http://the-internet.herokuapp.com/download')
  download_link = @driver.find_element(css: '.example a')
  download_link.click
  sleep 0.5 # If execution is too fast you can get the temporary downloaded file when you scrape the directory
  files = Dir.glob("#{@download_dir}/**")

  expect(files.size).to eq 1
  file = files.first

  expect(File.size(file)).to be > 0
end

File: 02-download-a-file/python/download.py
# -*- coding: utf-8 -*-
"""
Implementation of http://elementalselenium.com/tips/2-download-a-file
"""

import os
import time
import shutil
import tempfile
import unittest
from selenium import webdriver


class Download(unittest.TestCase):

    def setUp(self):
        self.download_dir = tempfile.mkdtemp()

        profile = webdriver.FirefoxProfile()
        profile.set_preference("browser.download.dir", self.download_dir)
        profile.set_preference("browser.download.folderList", 2)
        profile.set_preference(
            "browser.helperApps.neverAsk.saveToDisk",
            "images/jpeg, application/pdf, application/octet-stream")
        profile.set_preference("pdfjs.disabled", True)
        self.driver = webdriver.Firefox(firefox_profile=profile)

    def tearDown(self):
        self.driver.quit()
        shutil.rmtree(self.download_dir)

    def test_example_1(self):
        driver = self.driver
        driver.get('http://the-internet.herokuapp.com/download')
        download_link = driver.find_element_by_css_selector('.example a')
        download_link.click()

        time.sleep(1.0)  # necessary for slow download speeds

        files = os.listdir(self.download_dir)
        files = [os.path.join(self.download_dir, f)
                 for f in files]  # add directory to each filename
        assert len(files) > 0, "no files were downloaded"
        assert os.path.getsize(files[0]) > 0, "downloaded file was empty"

if __name__ == "__main__":
    unittest.main()
